src/uct/ib/dc/dc_mlx5_ep.h

```diff
@@ -352,12 +352,22 @@ uct_dc_mlx5_iface_progress_pending(uct_dc_mlx5_iface_t *iface,
                           uct_dc_mlx5_iface_dci_can_alloc(iface, pool_index)));
 }
 
-static inline int uct_dc_mlx5_iface_dci_ep_can_send(uct_dc_mlx5_ep_t *ep)
+static inline int uct_dc_mlx5_iface_dci_ep_can_send(uct_dc_mlx5_ep_t *ep,
+                                                    int flush_cancel)
 {
-    uct_dc_mlx5_iface_t *iface = ucs_derived_of(ep->super.super.iface, uct_dc_mlx5_iface_t);
-    return (!(ep->flags & UCT_DC_MLX5_EP_FLAG_TX_WAIT)) &&
-           uct_rc_fc_has_resources(&iface->super.super, &ep->fc) &&
-           uct_dc_mlx5_iface_dci_has_tx_resources(iface, ep->dci);
+    uct_dc_mlx5_iface_t *iface = ucs_derived_of(ep->super.super.iface,
+                                                uct_dc_mlx5_iface_t);
+
+    if (!uct_dc_mlx5_iface_dci_has_tx_resources(iface, ep->dci)) {
+        return 0;
+    } if (flush_cancel) {
```

Comment:
adding this flag "flush_cancel" is not right way: it creates an opportunity to break ordering between flush and other operations that were pending waiting for FC window,
why not enough to reset FC window after all completions (with error) have arrived?

Reply:
> why not enough to reset FC window after all completions (with error) have arrived?

yes, it is enough

> adding this flag "flush_cancel" is not right way: it creates an opportunity to break ordering between flush and other operations that were pending waiting for FC window

it aligns RC and DC behavior

Reply:
but we could remove it from this PR and think twice which transport should be fixed - RC or DC

Reply:
can we reset FC state in both RC and DC error handlers (after ci==pi)?

Reply:
> can we reset FC state in both RC and DC error handlers (after ci==pi)?

yes, we can. but anyway user should post nothing (except NOP in `flush(CANCEL)`) if error was detected.
do you see any problem with not waiting for `ci==pi`?

Reply:
actually, I don't see any problem updating FC window if `ci < pi`. if we have pending operations they will be purged by a user in error handling callback (e.g. as UCP does); otherwise - the next TX operation will fail an assertion that prevents sending data on failed UCT EPs.

Reply:
So do it the first time we detect a failure?

Reply:
> So do it the first time we detect a failure?

do you mean ensure resetting FC window size only once? unfortunately, we don't have counter/flag (have only debug-only flag on TXWQ) which indicates if the error was detected or not.

Reply:
we must have something to avoid calling user err callback more than once from UCT

Reply:
> we must have something to avoid calling user err callback more than once from UCT

yes, we have flags, but what if user's error callback wasn't set?
also, why do we need to set it only once? if no TX operations will be done, no one will decrease FC window size

Reply:
we have this for dc:
```
    if (ep->flags & (UCT_DC_MLX5_EP_FLAG_ERR_HANDLER_INVOKED |
                     UCT_DC_MLX5_EP_FLAG_FLUSH_CANCEL)) {
        return;
    }
```

and this for rc:
```
    if (ep->super.flags & (UCT_RC_EP_FLAG_ERR_HANDLER_INVOKED |
                           UCT_RC_EP_FLAG_FLUSH_CANCEL)) {
        goto out;
    }
```

so i guess it should be ok

Reply:
yes, we could use it.
but `UCT_RC_UPDATE_FC_WND` is called when doing some TX operations which shouldn't be done in case of UCT EP failed.
wdyt about adding `assert(!((_ep)->super.flags & UCT_RC_EP_FLAG_ERR_HANDLER_INVOKED))` inside `UCT_RC_UPDATE_FC_WND`?

Reply:
hmm, it is common code for both RC and DC. it will require adding wrapper for DC to check `ERR_HANDLER_INVOKED` flag and use already existing `UCT_RC_UPDATE_FC` wrapper for RC.

Reply:
i don't like adding asserts in unrelated places
imo the logically expected flow is to just reset fc once when ep has failed

Reply:
ok, done